### PR TITLE
Don't use Vector<byte>.Dot

### LINKED ISSUE #493: Don't use Vector<byte>.Dot
`Vector<byte>.Dot` is not an intrinsic so quite slow; see https://github.com/dotnet/coreclr/issues/2359 so don't use it. 

> `Vector.Dot<byte>` is [not an intrinsic](https://github.com/dotnet/coreclr/blob/master/src/jit/simd.cpp#L2392) single CPU instruction and is instead composed of a [new object create] (https://github.com/dotnet/corefx/blob/master/src/System.Numerics.Vectors/src/System/Numerics/Vector.cs#L4142) followed by [32 byte multiply instructions and 32 byte addition instructions](https://github.com/dotnet/corefx/blob/master/src/System.Numerics.Vectors/src/System/Numerics/Vector.cs#L4145) (depending on arch). 
> dotproduct-cpu
> For the aspnet5 server Kestrel this adds up to being 8.23% of the *total* server CPU load as it makes heavy use of it when parsing the ASCII headers.

Need alternative implementation for `MemoryPoolIterator2.Seek`

### PR DESCRIPTION
**Pull request by benaadams**
23% improvement in RPS

Faster implementation than using `Vector<byte>.Dot`

Resolves #493

Before
> Running 30s test @ http://.../plaintext
  32 threads and 1024 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     8.68ms   14.73ms 333.56ms   95.18%
    Req/Sec    47.81k     4.58k  122.10k    90.28%
  44317883 requests in 30.10s, 5.45GB read
  Socket errors: connect 35, read 0, write 0, timeout 0
Requests/sec: 1472388.96

After
> Running 30s test @ http://.../plaintext
  32 threads and 1024 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency    11.18ms   28.91ms 558.18ms   94.56%
    Req/Sec    59.52k     8.66k  282.83k    93.10%
  54796839 requests in 30.10s, 6.74GB read
  Socket errors: connect 35, read 0, write 0, timeout 0
Requests/sec: 1820474.06

### CODE REVIEW

**Comment by benaadams** [comment]
Has a bug - adding more tests

**Comment by halter73** [comment]
23%! That's great. We're going to run this in our perflab along with the StringValues change to a class (https://github.com/aspnet/Common/pull/68) and report back with the combined results.

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs [hunk 1]
```diff
@@ -146,7 +133,6 @@ public int Seek(int char0)
             }
 
             var byte0 = (byte)char0;
-            var vectorStride = Vector<byte>.Count;
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: why this change?

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: Read that it's Jitted to a const; will try to find reference

File: src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs [hunk 2]
```diff
@@ -417,6 +397,34 @@ public int Seek(int char0, int char1, int char2)
...
+            // Quasi-tree search
+            var vector64 = Vector.AsVectorInt64(chEquals);
+            for (var i = 0; i < Vector<long>.Count; i++)
+            {
+                var longValue = vector64[i];
+                if (longValue == 0) continue;
+
+                var shift = i << 1;
+                var offset = shift << 2;
+                var vector32 = Vector.AsVectorInt32(chEquals);
```
Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: Out of curiosity, have you done any micro benchmarks to confirm that creating `vector32` to determine whether or not to check the first or second half of the long is faster than naively checking up to all 8 bytes for random/evenly distributed `chEquals` inputs?

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: Is probably worth a test; though the `AsVectorType` conversions shouldn't* do anything post Jit in release as its all the same register and should just be changing the cpu instructions called on them. Also one of the reasons I did it with vector indexers `[]` rather than shift+cast; is endianness confuses me.

*doesn't mean they don't

Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: Cool. I'm going to try to get this in quickly. Hard to argue with a 23% RPS gain!

Comment (benaadams) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: I can do a follow up PR if need be, after testing. Might be some stuff can get moved further down into coreclr/corefx types.

Comment (halter73) [inline src/Microsoft.AspNet.Server.Kestrel/Infrastructure/MemoryPoolIterator2.cs]: I did the microbenchmarking myself ([Program.cs](https://gist.github.com/halter73/236f9f300b7f5cfe1d40)), and it appears that it's better to do the "Quasi-tree search". At least it doesn't seem any worse. Here are the results from my laptop:

    C:\Users\shalter\dev\Universe\KestrelHttpServer\samples\ConsoleApp1 [dev +1 ~2 -0 !]> # FindFirstEqualByte
    C:\Users\shalter\dev\Universe\KestrelHttpServer\samples\ConsoleApp1 [dev +1 ~2 -0 !]> dnx --configuration Release ConsoleApp1
    1635
    C:\Users\shalter\dev\Universe\KestrelHttpServer\samples\ConsoleApp1 [dev +1 ~2 -0 !]> dnx --configuration Release ConsoleApp1
    1640
    C:\Users\shalter\dev\Universe\KestrelHttpServer\samples\ConsoleApp1 [dev +1 ~2 -0 !]> dnx --configuration Release ConsoleApp1
    1453

    C:\Users\shalter\dev\Universe\KestrelHttpServer\samples\ConsoleApp1 [dev +1 ~2 -0 !]> # FindFirstEqualByte2
    C:\Users\shalter\dev\Universe\KestrelHttpServer\samples\ConsoleApp1 [dev +1 ~2 -0 !]> dnx --configuration Release ConsoleApp1
    1715
    C:\Users\shalter\dev\Universe\KestrelHttpServer\samples\ConsoleApp1 [dev +1 ~2 -0 !]> dnx --configuration Release ConsoleApp1
    1697
    C:\Users\shalter\dev\Universe\KestrelHttpServer\samples\ConsoleApp1 [dev +1 ~2 -0 !]> dnx --configuration Release ConsoleApp1
    1688

Merging now.

**Comment by darcythomas** [comment]
What is the environment that you ran these tests on? (So I can make some apple to apple comparisons)

**Comment by halter73** [comment]
1.0.0-rc2-16312 coreclr x64 win dnx on my laptop (intel core i7-3367U)

**Comment by halter73** [comment]
It's hard to make apple-to-apple comparisons unless you run the baseline on the exact same machine.

**Comment by darcythomas** [comment]
Apple to pear comparisons perhaps :P 

I was wondering if it was something "Standard-ish" like a A3 VM instance on Azure, or something...

**Comment by benaadams** [comment]
@darcythomas you need to run in release (SIMD optimizations are weakened in debug; so its possible to debug them). `Vector<float>.Dot` is hardware accelerated, but there isn't an equivalent cpu instruction for `Vector<byte>.Dot`.

On a A3 VM you might max out its network first; I'm testing with G4 VMs as already above 2.3GBps sends.
